src/artifacts.js

```diff
@@ -254,23 +325,85 @@ api.declare({
         });
     }
 
+    if (artifact.storageType === 'blob') {
+      if (artifact.details.uploadId) {
+        // TODO: Here's where we return a new set of signed requests to do
+        // with this.s3Controller.genereate....();
+      }
+    }
+
     // Update expiration and detail, which may have been modified
     await artifact.modify((artifact) => {
       artifact.expires = expires;
       artifact.details = details;
     });
   }
 
-  // Publish message about artifact creation
-  await this.publisher.artifactCreated({
-    status:         task.status(),
-    artifact:       artifact.json(),
-    workerGroup,
-    workerId,
-    runId,
-  }, task.routes);
+
+  // This event is *invalid* for s3/azure storage types so we'll stop sending it.
+  // It's only valid for error, reference and blob
+  if (artifact.storageType === 'error' || artifact.storageType === 'reference') {
+    // Publish message about artifact creation
+    await this.publisher.artifactCreated({
+      status:         task.status(),
+      artifact:       artifact.json(),
+      workerGroup,
+      workerId,
+      runId,
+    }, task.routes);
+  }
 
   switch (artifact.storageType) {
+    case 'blob':
+      let requests;
+      let uploadId;
+      // If we're supposed to do a multipart upload, we should generate an UploadId
+      // if it doesn't already exist.  We should store that ID in the entity
+      if (artifact.details.parts) {
+        if (!artifact.details.uploadId) {
+          uploadId = await this.s3Controller.initiateMultipartUpload({
+            bucket: artifact.details.bucket,
+            key: artifact.details.key,
+            sha256: artifact.details.sha256,
+            size: artifact.details.size,
+            metadata: {taskId, runId, name},
+            tags: {taskId, runId, name},
+            contentType: artifact.contentType,
+            contentEncoding: artifact.details.encoding || undefined,
+          });
+          await artifact.modify(_artifact => {
+            _artifact.details.uploadId = uploadId;
```

Comment:
This is a a bit complicated... Particularly, handling the `_.isEqual` part of the create entity flow...

But if we're lucky it might just turn up somewhat elegant in the end :)
I mean cancelling the uploadId if there is a conflict might not be a hard I thought...